### DISCUSSION TITLE
Removal of Oddity?

### DISCUSSION THREAD

**Discussion by nguyen-andrew** [emichael/dslabs#39] [General]
Hi Ellis, thanks for creating this framework and sharing it - it's been an invaluable learning resource. I noticed that the labs recently removed Oddity and switched to a new visualizer. Can I ask what motivated you to make the switch? I'm interested in what the new visualizer can do, but I'm also curious about what was missing in the Oddity visualizer. Thanks!

**Comment by emichael** [comment]
The main reason was that Oddity was written in Clojurescript. Working on Oddity itself was difficult. Also, all communication between Oddity and DSLabs had to happen in JSON or some other serialized format. One way in which that was annoying was that objects with circular references couldn't be handled without a custom serializer.

Now, the visualizer and DSLabs tests are run out of the same JVM, and it's surpassed Oddity in terms of the features provided. My favorite aspect is the event list on the right hand side of the display. It provides a graphical and expandable view of the messages/timers that lead to the current state. When I'm debugging, looking at this is often the fastest way to diagnose problems.

It should be easy enough to try it yourself. All the same commands to `run-tests.py` still work. You can use the example in lab 0 to test it out.

The one downside of the new visualizer is that running a Java GUI app isn't quite as foolproof as opening up a web browser in terms of cross-platform compatibility, but I've been able to run it on all the major operating systems without issue.

**Reply by nguyen-andrew** [reply]
Thanks, that makes a lot of sense. It was pretty straightforward to start like you said. I did run into an issue with the GUI's background not displaying correctly as shown in the image below, but I found out it was more of an issue with my specific setup (M1 Mac, docker, & XQuartz) than with the visualizer itself.

![image](https://user-images.githubusercontent.com/37893562/192654594-f871394d-2b0d-4e2a-9627-698079d130a1.png)
(Took a photo of my screen because screenshots didn't capture the issue.)

I discovered from [this issue thread](https://github.com/XQuartz/XQuartz/issues/31#issuecomment-1181028344) that I could fix it by adding the following to `run_viz_debugger` of the `run-tests.py` in the handout:
```python
    command.append('-Dsun.java2d.xrender=false')
    command.append('-Dsun.java2d.pmoffscreen=false')
```

Anyways, I'm excited to poke around with the new visualizer. I do think that the collapsable windows in Oddity were nice, and they made it quick & easy to inspect an entity and see all of its state. I think a similar mechanism would be great here as well instead of having to scroll to see all of the information in each of the panels.

**Reply by emichael** [reply]
Re: visual glitches, please feel free to open up an issue! I want to fix these issues and have already done some hacky things on startup to include platform-specific fixes.

Re: state inspection, everything is expandable! Just double-click. Maybe there should be better docs or maybe the UI should make that bit of feature discovery more obvious.

**Reply by emichael** [reply]
@andrewnguyen92 if you comment out these lines in `DebuggerWindow.java`
```
        if (!runningInWSL()) {
            System.setProperty("sun.java2d.opengl", "true");
        }
        System.setProperty("sun.java2d.noddraw", "true");
```
but don't do anything else like appending those extra parameters in `run-tests.py`, do you still get those visual glitches? I'm thinking at this point that trying to enable opengl is more trouble than it's worth.

**Reply by nguyen-andrew** [reply]
That sounds good - I'm not sure how many other people are running dslabs with a similar setup as mine, but I'll write up an issue and post my fix in case anyone is. And yea I still get that glitch even with those lines commented out.
